nodejs/aws-infra/experimental/containerDefinition.ts

```diff
@@ -0,0 +1,153 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as aws from "@pulumi/aws";
+import * as pulumi from "@pulumi/pulumi";
+
+import * as mod from ".";
+
+import * as utils from "../utils";
+
+export type WrappedEndpoints = Record<string, Record<number, pulumi.Output<aws.apigateway.x.Endpoint>>>;
+
+/** @internal */
+export function computeContainerDefinition(
+    parent: pulumi.Resource,
+    name: string,
+    containerName: string,
+    container: ContainerDefinition,
+    logGroup: aws.cloudwatch.LogGroup): pulumi.Output<aws.ecs.ContainerDefinition> {
+
+    if (container.image === undefined && container.imageProvider === undefined) {
+        throw new Error("container requires either [image] or [imageProvider] to be set.");
+    }
+
+    const imageProvider = container.imageProvider;
+    const loadBalancerProvider = container.loadBalancerProvider;
+
+    let image = container.image;
+    let environment = container.environment;
+    let portMappings = container.portMappings;
+
+    if (imageProvider) {
+        image = imageProvider.image(name, parent);
+        environment = utils.combineArrays(
+            environment, imageProvider.environment(name, parent));
+    }
+
+    if (loadBalancerProvider) {
+        portMappings = utils.combineArrays(
+            portMappings, loadBalancerProvider.portMappings(containerName, name, parent));
+    }
```

Comment:
image, environment, and portMappings can all be user provided, but they all can be augmented by providers as well.  for single-data properties (like `image`), they provider can just supply the value.  For multi-valued properties (like `environment` or `portMappings`) the provider can add values.